src/app/util/attribute-storage.cpp

```diff
@@ -266,9 +266,64 @@ uint16_t emberAfGetDynamicIndexFromEndpoint(EndpointId id)
     return kEmberInvalidEndpointIndex;
 }
 
+namespace {
+
+const EmberAfCluster * getClusterTypeDefinition(EndpointId endpointId, ClusterId clusterId, EmberAfClusterMask mask)
+{
+    uint16_t index = emberAfIndexFromEndpointIncludingDisabledEndpoints(endpointId);
+    if (index != kEmberInvalidEndpointIndex)
+    {
+        return emberAfFindClusterInType(emAfEndpoints[index].endpointType, clusterId, mask, nullptr);
+    }
+    // not found
+    return nullptr;
+}
+
+} // anonymous namespace
+
+void emberAfSetupDynamicEndpointDeclaration(EmberAfEndpointType & endpointType, EndpointId templateEndpointId,
+                                            const Span<const EmberAfClusterSpec> & templateClusterSpecs)
+{
+    // allocate cluster list
+    size_t clusterCount = templateClusterSpecs.size();
+    VerifyOrDieWithMsg(clusterCount <= 255, Support, "max 255 clusters per endpoint!");
+    endpointType.clusterCount = static_cast<uint8_t>(clusterCount);
+    endpointType.cluster      = new EmberAfCluster[endpointType.clusterCount];
+    endpointType.endpointSize = 0;
+    // get the actual cluster pointers and sum up memory size
+    for (size_t i = 0; i < templateClusterSpecs.size(); i++)
+    {
+        auto cluster =
+            getClusterTypeDefinition(templateEndpointId, templateClusterSpecs[i].clusterId, templateClusterSpecs[i].mask);
+        VerifyOrDieWithMsg(cluster, Support, "cluster 0x%04x with mask %x could not be found in template endpoint %u",
+                           (unsigned int) templateClusterSpecs[i].clusterId, templateClusterSpecs[i].mask,
+                           (unsigned int) templateEndpointId);
+        // for now, we need to copy the cluster definition, unfortunately.
+        // TODO: make endpointType use a pointer to a list of EmberAfCluster* instead, so we can re-use cluster definitions
+        //   instead of duplicating them here once for every instance.
+        // Note: we need const cast here, because cluster in EmberAfEndpointType is const due
+        //   to the way static cluster code generation works.
+        const_cast<EmberAfCluster &>(endpointType.cluster[i]) = *cluster;
```

Comment:
Please don't const cast. Place the array of clusters in a local variable and move the assignment of `endpointType.cluster = localArray` after the for loop.

Reply:
I don't see how that would solve the problem. I can't create that array other than with `new EmberAfCluster[]`, and so I'm inheriting the const members of that, which prevent that assignment without the const cast. Am I missing something?